src/main/java/com/ncc/aif/ValidateAIF.java

```diff
@@ -574,4 +592,183 @@ public Resource validateKBAndReturnReport(Model dataToBeValidated, Model union)
     public static boolean isValidReport(Resource validationReport) {
         return validationReport.getRequiredProperty(CONFORMS).getBoolean();
     }
+
+
+    /**
+     * A statistics collector for use in profiling TopBraid-based SHACL validation.  Typical usage is to call
+     * {@link #startCollection()} and {@link #endCollection()} to bound statistics collection,
+     * then call {@link #dump(String)} to dump slow query statistics to <filename>-stats.txt.
+     *
+     * This statistics collector only outputs slow queries via the {@link #dump(String)} method.  If you suspect
+     * validation will not complete due to out of memory or other error conditions, consider using {@link ProgressiveStatsCollector}.
+     */
+    private static class StatsCollector {
+
+        private final int durationThreshold;
+
+        /**
+         * Creates a statistics collector that saves queries slower than [threshold] ms to a file.
+         * @param threshold the threshold definition of a slow query for this statistics collector
+         */
+        StatsCollector(int threshold) {
+            this.durationThreshold = threshold;
+        }
+
+        /**
+         * Start statistics collection.  Clears any previous statistics gathered by TopBraid statistics manager.
+         */
+        void startCollection() {
+            ExecStatisticsManager.get().reset();
+            ExecStatisticsManager.get().setRecording(true);
+        }
+
+        /**
+         * End statistics collection.
+         */
+        void endCollection() {
+            ExecStatisticsManager.get().setRecording(false);
+        }
+
+        /**
+         * Dump all gathered slow query statistics to <basename>-stats.txt, starting with the slowest queries.
+         * @param basename a file basename to determine the profiling output filename
+         */
+        void dump(String basename) {
+            final String outputFilename = basename.replace(".ttl", "-stats.txt");
+            try {
+                final PrintStream out = new PrintStream(java.nio.file.Files.newOutputStream(Paths.get(outputFilename)));
+                dumpStats(out);
+                out.close();
+            }
+            catch (IOException ioe) {
+                logger.warn("---> Could not write statistics for " + basename + ".");
+            }
+        }
+
+        // Dump stats to the specified PrintStream
+        private void dumpStats(PrintStream out) {
+            final SortedMap<Integer, ExecStatistics> savedStats = new TreeMap<>();
+            final SortedSet<Map.Entry<Integer, ExecStatistics>> sortedStats = new TreeSet<>(
+                    (e1, e2) -> Long.compare(e2.getValue().getDuration(), e1.getValue().getDuration()));
```

Comment:
I believe a Comparator will sort them in ascending order, but I want them descending.  That said, I just noticed Comparator.reversed().  That looks promising; I'll give it a look.

Reply:
I have not been able to get a `Comparator` to initialize a `TreeSet` this way and sort them in reverse order.  There may be other approaches using Collections or streams or something.  Somebody is welcome to give it a shot-- I'd be happy to incorporate it.

Reply:
I'm not sure it's even worth it anymore, but here you go:
```suggestion
                    Collections.reverseOrder(Comparator.comparing(entry -> entry.getValue().getDuration()))
```

Reply:
Well, I'd argue it's a bit more readable, at the cost of an O(n) operation.

Reply:
Haha, if you say so. Given that it's just as long as what you initially wrote, I'm inclined to go with your initial solution.

Does it cost an O(n)? I think it's just replacing compare (c1, c2) with (c2, c1). Let me know if I'm missing something

Reply:
Given that it's more readable, I actually prefer your solution, regardless of relative length.  It more clearly shows that you're reversing the order, whereas the original code did that by switching e2 with e1, which is more easily missed.  That said, the same readability could be achieved with a comment.  :)

The original code does a comparison of two longs in creating the `TreeSet`.  My take is that the new code does a (slightly different) comparison to create a `TreeSet`, then uses `Collections.reverseOrder()` to reverse the order of the items in the resulting collection.  I'm guessing `Collections.reverseOrder( )` is an O(n) operation, unless it's smart enough to optimize the code behind the scenes and just reverse the comparison.

Reply:
Either way, @sharknc, I think I've incorporated all of the review comments.  Let me know if I missed something or botched something.   :)

Reply:
Sounds good, I'll take a look and add any additional comments or merge tonight.

Reply:
@nextcen-dgemoets, I _think_ it modifies the comparator, which is why you don't have to call reverse() after every addition to the collection. Maybe we can discuss it next time your in the office because I'm still not 100% sure how it does it.